bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryTableOperations.java

```diff
@@ -0,0 +1,300 @@
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.CommitStateUnknownException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.hadoop.ConfigProperties;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Handles BigQuery metastore table operations. */
+public final class BigQueryTableOperations extends BaseMetastoreTableOperations {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryTableOperations.class);
+
+  public static final String TABLE_PROPERTIES_BQ_CONNECTION = "bq_connection";
+
+  private final BigQueryMetaStoreClient client;
+  private final FileIO fileIO;
+  private final TableReference tableReference;
+  private final Configuration conf;
+
+  BigQueryTableOperations(
+      BigQueryMetaStoreClient client,
+      FileIO fileIO,
+      String project,
+      String dataset,
+      String table,
+      Configuration conf) {
+    this.client = client;
+    this.fileIO = fileIO;
+    this.tableReference =
+        new TableReference().setProjectId(project).setDatasetId(dataset).setTableId(table);
+    this.conf = conf;
+  }
+
+  // The doRefresh method should provide implementation on how to get the metadata location.
+  @Override
+  public void doRefresh() {
+    // Must default to null.
+    String metadataLocation = null;
+    try {
+      metadataLocation =
+          getMetadataLocationOrThrow(
+              client.getTable(this.tableReference).getExternalCatalogTableOptions());
+    } catch (NoSuchTableException e) {
+      if (currentMetadataLocation() != null) {
+        // Re-throws the exception because the table must exist in this case.
+        throw e;
+      }
+    }
+    refreshFromMetadataLocation(metadataLocation);
+  }
+
+  // The doCommit method should provide implementation on how to update with metadata location
+  // atomically
+  @Override
+  public void doCommit(TableMetadata base, TableMetadata metadata) {
+    String newMetadataLocation =
+        base == null && metadata.metadataFileLocation() != null
+            ? metadata.metadataFileLocation()
+            : writeNewMetadata(metadata, currentVersion() + 1);
+    BaseMetastoreOperations.CommitStatus commitStatus =
+        BaseMetastoreOperations.CommitStatus.FAILURE;
+    try {
+      if (base == null) {
+        createTable(newMetadataLocation, metadata);
+      } else {
+        updateTable(base.metadataFileLocation(), newMetadataLocation, metadata);
+      }
+      commitStatus = BaseMetastoreOperations.CommitStatus.SUCCESS;
+    } catch (CommitFailedException | CommitStateUnknownException e) {
+      throw e;
+    } catch (Throwable e) {
+      LOG.error("Exception thrown on commit: ", e);
+
+      if (commitStatus == BaseMetastoreOperations.CommitStatus.FAILURE
```

Comment:
`commitStatus == BaseMetastoreOperations.CommitStatus.FAILURE` is always true in the current logic.